Let lighteval support sglang

Jayon02 opened the pull request:
You can use sglang in lighteval tasks

`lighteval sglang \
    "pretrained=HuggingFaceH4/zephyr-7b-beta,dtype=float16" \
    "helm|bigbench:bbq_lite_json:age_disambig|0|0"`

NathanHB commented on src/lighteval/models/vllm/vllm_model.py:
```diff
@@ -48,7 +48,6 @@
 
 logger = logging.getLogger(__name__)
 
-
```
this file should not be modified in the PR

NathanHB commented on src/lighteval/tasks/registry.py:
```diff
@@ -299,6 +299,7 @@ def taskinfo_selector(tasks: str, task_registry: Registry) -> tuple[list[str], d
         expanded_tasks = task_registry.task_groups_dict.get(maybe_task_group, [maybe_task_group])
         expanded_tasks_list.extend(expanded_tasks)
 
+    ## task expand and few shot number record, not load task
```
what do you mean in this comment ?

NathanHB commented on src/lighteval/main_vllm.py:
```diff
@@ -137,6 +137,7 @@ def vllm(
         model_config = VLLMModelConfig(config, generation_parameters=generation_parameters)
 
     else:
+        ## cmd arg
```
not needed

Jayon02 replied:
I'm sorry for having a bad development habit. I'll be more mindful in my upcoming commits.

NathanHB replied:
it's completely fine ! if it helps you it's ok, I'm just adding comments so we don't forget to remove them once merging :)

NathanHB commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,295 @@
+    ) -> list[GenerativeResponse]:
+        """Contains the actual logic of the generation."""
+        # TODO: double check
+        self.sampling_params["stop"] = stop_tokens
+        self.sampling_params["n"] = num_samples
+        self.sampling_params["top_p"] = 1.0
+        self.sampling_params["top_k"] = -1
+        self.sampling_params["skip_special_tokens"] = True
+        self.sampling_params["temperature"] = 0
```
why are you forcing those configs ?

zhaochenyang20 replied:
Yeah. Maybe we should change this:

https://github.com/huggingface/lighteval/pull/552#discussion_r1953341357

But when you are testing, you should force this.

NathanHB commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,295 @@
+
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
+
+STARTING_BATCH_SIZE = 512
```
is this needed for sglang ?

zhaochenyang20 replied:
I think not really. Can only use the defualt parameters.

zhaochenyang20 commented on src/lighteval/main_sglang.py:
```diff
@@ -0,0 +1,158 @@
+# MIT License
+
+# Copyright (c) 2024 The HuggingFace Team
```
We can say this is contributed by Qiujiang Chen from SGLang team also. 😂

NathanHB replied:
Of course !

zhaochenyang20 commented on src/lighteval/main_sglang.py:
```diff
@@ -0,0 +1,158 @@
+
+    results = pipeline.get_results()
+
+    pipeline.save_and_push_results()
```
Do we need to close the engine here? Or it's already in the `pipeline`?

Jayon02 replied:
The engine is closed in `pipeline.evaluate()`

zhaochenyang20 commented on src/lighteval/models/model_input.py:
```diff
@@ -119,3 +119,12 @@ def to_tgi_ie_dict(self) -> dict:
         }
         return {k: v for k, v in args.items() if v is not None}
+
+    def to_sglang_dict(self) -> dict:
+        args = {
+            "max_new_tokens": self.max_new_tokens,
+            "stop_token_ids": self.stop_tokens,
+            "temperature": self.temperature,
+        }
+        return {k: v for k, v in args.items() if v is not None}
```
This style looks okay. But do we only have these parameters? TGI has something like `top_p` `seed`, these actually also in SGLang. You can see this:

https://docs.sglang.ai/references/sampling_params.html

Also, this docs are still being developing by simon (Simon V) in our SGLang slack. You can ask him for help also.

zhaochenyang20 replied:
> This style looks okay. But do we only have these parameters? TGI has something like `top_p` `seed`, these actually also in SGLang. You can see this:
> 
> https://docs.sglang.ai/references/sampling_params.html
> 
> Also, this docs are still being developing by simon (Simon V) in our SGLang slack. You can ask him for help also.

Also, you can see this PR for reference:

https://github.com/sgl-project/sglang/pull/3260

zhaochenyang20 commented on src/lighteval/models/model_loader.py:
```diff
@@ -43,7 +44,7 @@
     is_openai_available,
     is_tgi_available,
-    is_vllm_available,
+    is_vllm_available, is_sglang_available, NO_SGLANG_ERROR_MSG,
```
obviously you should use `black` to stylify this. Check how SGLang pre-commit works and learn to use `black` for python. Thnask!

clefourrier replied:
We don't use black but ruff - please check out  the pre-commit hook of lighteval

Jayon02 replied:
I have used pre-commit to fix all issues

zhaochenyang20 commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,295 @@
+if is_sglang_available():
+    from sglang import Engine
+    from sglang.srt.hf_transformers_utils import get_tokenizer
+    from sglang.srt.distributed.parallel_state import destroy_model_parallel, destroy_distributed_environment
```
Actually, if you are using the Engine, destroy an engine is pretty easy. Just `Engine.shut_down` like our docs. (I don't remember the full name)

Why we need `destroy_model_parallel`, `destroy_distributed_environment`.

zhaochenyang20 commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,295 @@
+STARTING_BATCH_SIZE = 512
+
+@dataclass
+class SGLANGModelConfig:
```
Do not use `SGLANGModelConfig` this kind of name. Rather using `SGLangModelConfig`. Only captilize SGL, leave `ang`.

zhaochenyang20 commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,295 @@
+        if not self.generation_parameters:
+            self.generation_parameters = GenerationParameters()
+
+class SGLANGModel(LightevalModel):
```
change the name also.

zhaochenyang20 commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,295 @@
+        self._max_length = int(config.context_length) if config.context_length is not None else None
+        self.model = self._create_auto_model(config, env_config)
+        self.model_name = _simplify_name(config.pretrained)
+        self.model_sha = ""  # config.get_model_sha()
```
what is `sha`?

clefourrier replied:
the latest commit of the current model, if you're grabbing it from the web

zhaochenyang20 commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,295 @@
+            "log_level": "info",
+        }
+
+        if config.dp_size > 1:
```
Actually we can set this. But need to test it carefully.

zhaochenyang20 commented on src/lighteval/pipeline.py:
```diff
@@ -57,7 +57,7 @@
     is_openai_available,
     is_tgi_available,
-    is_vllm_available,
+    is_vllm_available, is_sglang_available, NO_SGLANG_ERROR_MSG,
```
use black to lint it

Jayon02 commented:
1. I have addressed all the issues mentioned above. I apologize for the problems that occurred during my development process.
2. I add sglang as a potential backend for lighteval, offering high accuracy and efficiency.
3. I provide an [initial document](https://github.com/Jayon02/lighteval/blob/main/docs/source/use-sglang-as-backend.mdx) describing how to use sglang in lighteval after [installing sglang](https://docs.sglang.ai/start/install.html) in your device .
4. Here is an example of how to run lighteval with sglang, including adding model arguments and sampling arguments.
```
lighteval sglang \
"pretrained=Qwen/Qwen2-7B,dtype=float16,tp_size=2,generation_parameters={temperature: 1.0}" \
"leaderboard|truthfulqa:mc|0|0"
```
Please let me know if there is anything that could be improved. Thanks to all SGLang team members for their help.

Jayon02 commented:
I conduct some experiments to compare the metrics of lighteval using sglang and vllm.

The experimental setup is as follows:
```
NVIDIA A100
Linux 5.15.0-126-generic
lighteval                         0.6.0.dev0
python                            3.13.0
torch                             2.5.1
vllm                              0.7.2
sglang                            0.4.2.post4
sgl-kernel                        0.0.3.post3
flashinfer-python                 0.2.0.post2+cu124torch2.5
```
However, I find that the parameter **temperature** is set to 1.0 in some vllm tests. This parameter increases the uncertainty of the model's generated results. Here are some results and t means temperature below:

  | metric | sglang value (t = 1.0) | vllm value  (t = 1.0) | sglang value  (t = 0.0) | vllm value (t = 0.0)
-- | -- | -- | -- | -- | --
helm:truthfulqa | em | 0.5306 | 0.5046 | 0.6346 | 0.6346
  | qem | 0.5367 | 0.5122 | 0.6346 | 0.6346
  | pem | 0.5321 | 0.5061 | 0.6346 | 0.6346
  | pqem | 0.6223 | 0.6070 | 0.7080 | 0.7080
  | acc | 0.2584 | 0.2584 | 0.2584 | 0.2584
helm:siqa | em | 0.1479 | 0.1428 | 0.1018 | 0.1008
  | qem | 0.1535 | 0.1515 | 0.1018 | 0.1008
  | pem | 0.1515 | 0.1469 | 0.1018 | 0.1008
  | pqem | 0.4135 | 0.4186 | 0.3649 | 0.3644
harness:bbh:hyperbaton | em | 0.092 | 0.080 | 0 | 0
  | qem | 0.108 | 0.108 | 0 | 0
  | pem | 0.504 | 0.420 | 0.824 | 0.824
  | pqem | 0.780 | 0.740 | 0.968 | 0.968
  | perfect_em | 0 | 0 | 0 | 0
helm:mmlu | em | 0.6373 | 0.6475 | 0.7110 | 0.7111
  | qem | 0.6375 | 0.6475 | 0.7110 | 0.7111
  | pem | 0.6373 | 0.6475 | 0.7110 | 0.7111
  | pqem | 0.7245 | 0.7339 | 0.7832 | 0.7832

I find that when **temperature = 0.0**, the difference between sglang and vllm is very small. I also independently launch vllm and sglang backend without lighteval. When **temperature = 1.0**, for the same input, the results generated by each backend are different each time. I observe that each backend selects one output from multiple available options, and both backends have the same set of possible outputs. When evaluating, I think we should consider all options, not just one. The results in the table when **temperature = 1.0** is inaccurate。

In other experiments that only require using `loglikelihood` to compute metrics, such as *lighteval:logiqa*, and *leaderboard:truthfulqa:mc*, **temperature** is set to 0.0 and the results are nearly the same for both sglang and vllm.

I believe that for model evaluation, a more deterministic parameter setting should be chosen to better assess the model. So I think **temperature = 0.0** is a more reasonable setting. 

I am very eager to know your thoughts on the consideration of this parameter setting.

zhaochenyang20 commented:
@Jayon02 great work!

NathanHB commented on docs/source/installation.mdx:
```diff
@@ -31,6 +33,7 @@ appropriate extras group.
 | adapters     | To evaluate adapters models (delta and peft)                              |
 | tensorboardX | To upload your results to tensorboard                                     |
 | vllm         | To use vllm as backend for inference                                      |
+| sglang       | To use sglang as backend for inference                                    |
```
You did not modify the `pyproject` file to reflect this extra

Jayon02 replied:
I link `sglang document` for user to install [here](https://github.com/Jayon02/lighteval/blob/main/docs/source/installation.mdx). SGLang has some limitation which can't be added by `pyproject`. If we add `sglang = [sglang>=0.4.2.post]`, `pip install lighteval[sglang]` can't install dependency correctly. So I hope user can install sglang with sglang document.

NathanHB replied:
Great ! Well in that case, it's not really an extra and I would not add it to the table

NathanHB commented on docs/source/use-sglang-as-backend.mdx:
```diff
@@ -0,0 +1,52 @@
+
+> [!WARNING]
+> In the case of OOM issues, you might need to reduce the context size of the
+> model as well as reduce the `mem_fraction_static` and `chunked_prefill_size` parameter.
```
Could you add an example showing how to change to generation paramaters ? It can be found in our `use vllm as bakcend` doc

Jayon02 replied:
I add `sglang_model_config.yaml` in directory.

Jayon02 replied:
I also unified [the parameter configuration methods of sglang](https://github.com/Jayon02/lighteval/blob/d8841a9be46ef661a65ade98f563de9ea0989000/src/lighteval/main_sglang.py#L133) and vllm and modified the previous way of handling command-line arguments in sglang.

NathanHB commented on src/lighteval/main_sglang.py:
```diff
@@ -0,0 +1,171 @@
+
+    else:
+        pattern = re.compile(r"(\w+)=(\{.*\}|[^,]+)")
+        matches = pattern.findall(model_args)
+        model_args_dict = {}
+        generation_params = None
+        for key, value in matches:
+            key = key.strip()
+            if key == "generation_parameters":
+                value = re.sub(r"(\w+):", r'"\1":', value)
+                value = json.loads(value)
```
why didn't you use the way it's done for other backends ?

NathanHB commented on src/lighteval/models/sglang/sglang_model.py:
```diff
@@ -0,0 +1,339 @@
+            # Would we rather truncate the prompt to allow generation to go to max_new_tokens, at the risk
+            # of losing some meaning, or have some generations that are exceedingly short?
+            # The choice we go for here is to avoid truncating the prompt if we can, since it
+            # should have been managed by the prompt creator/few shot manager if requested by the user.
```
sometimes, even in 0 shots the prompt is too long, then I think we should truncate the prompt and allow generation, since we also evaluate model on context length

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/lighteval/pr_552). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zhaochenyang20 commented:
@Jayon02 This PR is great. Could you resolve / close the conversation if you solved it?

zhaochenyang20 commented:
@Jayon02 @NathanHB This PR is approved. When can we merge it?

NathanHB commented:
Waiting for tests and should be ready !
